{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import re\n",
    "import glob\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>ibp</th>\n",
       "      <th>jbp</th>\n",
       "      <th>fq</th>\n",
       "      <th>fdr</th>\n",
       "      <th>rawc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69740000</td>\n",
       "      <td>69780000</td>\n",
       "      <td>0.780465</td>\n",
       "      <td>0.0</td>\n",
       "      <td>29.353172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69740000</td>\n",
       "      <td>69790000</td>\n",
       "      <td>0.370194</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.741314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69740000</td>\n",
       "      <td>69950000</td>\n",
       "      <td>0.068101</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.943521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69740000</td>\n",
       "      <td>71210000</td>\n",
       "      <td>0.035838</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.222382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69745000</td>\n",
       "      <td>69800000</td>\n",
       "      <td>0.531329</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.584558</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr       ibp       jbp        fq  fdr       rawc\n",
       "0  chr1  69740000  69780000  0.780465  0.0  29.353172\n",
       "1  chr1  69740000  69790000  0.370194  0.0  22.741314\n",
       "2  chr1  69740000  69950000  0.068101  0.0  12.943521\n",
       "3  chr1  69740000  71210000  0.035838  0.0  10.222382\n",
       "4  chr1  69745000  69800000  0.531329  0.0  25.584558"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"../Data/all_processed_HiC/K_all_hic_cont.csv.gz\"\n",
    "df = pd.read_csv(file_path)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_refined_GM = pd.read_csv(\"../Data/all_processed_HiC/GM_all_hic_cont.csv.gz\")\n",
    "\n",
    "# get the average value of 'fq' and 'fdr' for each group\n",
    "df_refined_GM = df_refined_GM.groupby([\"chr\", \"ibp\", \"jbp\"], as_index=False).agg(\n",
    "    {\"fq\": \"mean\", \"fdr\": \"mean\", \"rawc\": \"first\"}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>ibp</th>\n",
       "      <th>jbp</th>\n",
       "      <th>fq</th>\n",
       "      <th>fdr</th>\n",
       "      <th>rawc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3320000</td>\n",
       "      <td>3320000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>156.021600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3320000</td>\n",
       "      <td>3325000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>59.187775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3320000</td>\n",
       "      <td>3330000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>39.219370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3320000</td>\n",
       "      <td>3335000</td>\n",
       "      <td>0.998299</td>\n",
       "      <td>0.981670</td>\n",
       "      <td>24.210155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3320000</td>\n",
       "      <td>3340000</td>\n",
       "      <td>0.809792</td>\n",
       "      <td>0.855084</td>\n",
       "      <td>20.779170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr      ibp      jbp        fq       fdr        rawc\n",
       "0  chr1  3320000  3320000  1.000000  1.000000  156.021600\n",
       "1  chr1  3320000  3325000  1.000000  1.000000   59.187775\n",
       "2  chr1  3320000  3330000  1.000000  1.000000   39.219370\n",
       "3  chr1  3320000  3335000  0.998299  0.981670   24.210155\n",
       "4  chr1  3320000  3340000  0.809792  0.855084   20.779170"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_refined_GM.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Gene Data**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Begin</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100287102</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1 (pseudogene)</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>653635</td>\n",
       "      <td>WASP family homolog 7, pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>102466751</td>\n",
       "      <td>microRNA 6859-1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>107985730</td>\n",
       "      <td>MIR1302-2 host gene</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>1</td>\n",
       "      <td>29774</td>\n",
       "      <td>35418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100302278</td>\n",
       "      <td>microRNA 1302-2</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68320</th>\n",
       "      <td>100289087</td>\n",
       "      <td>testis specific protein Y-linked 10</td>\n",
       "      <td>TSPY10</td>\n",
       "      <td>Y</td>\n",
       "      <td>752540</td>\n",
       "      <td>755334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68321</th>\n",
       "      <td>642631</td>\n",
       "      <td>testis specific protein Y-linked 15, pseudogene</td>\n",
       "      <td>TSPY15P</td>\n",
       "      <td>Y</td>\n",
       "      <td>772716</td>\n",
       "      <td>775527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68322</th>\n",
       "      <td>64593</td>\n",
       "      <td>RNA binding motif protein Y-linked family 3 me...</td>\n",
       "      <td>RBMY3AP</td>\n",
       "      <td>Y</td>\n",
       "      <td>835375</td>\n",
       "      <td>839807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68323</th>\n",
       "      <td>100874288</td>\n",
       "      <td>testis specific protein Y-linked 25, pseudogene</td>\n",
       "      <td>TSPY25P</td>\n",
       "      <td>Y</td>\n",
       "      <td>848837</td>\n",
       "      <td>851006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68324</th>\n",
       "      <td>727969</td>\n",
       "      <td>testis specific protein Y-linked 16, pseudogene</td>\n",
       "      <td>TSPY16P</td>\n",
       "      <td>Y</td>\n",
       "      <td>864903</td>\n",
       "      <td>865415</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68325 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene ID                                               Name  \\\n",
       "0      100287102         DEAD/H-box helicase 11 like 1 (pseudogene)   \n",
       "1         653635                  WASP family homolog 7, pseudogene   \n",
       "2      102466751                                    microRNA 6859-1   \n",
       "3      107985730                                MIR1302-2 host gene   \n",
       "4      100302278                                    microRNA 1302-2   \n",
       "...          ...                                                ...   \n",
       "68320  100289087                testis specific protein Y-linked 10   \n",
       "68321     642631    testis specific protein Y-linked 15, pseudogene   \n",
       "68322      64593  RNA binding motif protein Y-linked family 3 me...   \n",
       "68323  100874288    testis specific protein Y-linked 25, pseudogene   \n",
       "68324     727969    testis specific protein Y-linked 16, pseudogene   \n",
       "\n",
       "            Symbol Chromosome   Begin     End  \n",
       "0          DDX11L1          1   11874   14409  \n",
       "1           WASH7P          1   14362   29370  \n",
       "2        MIR6859-1          1   17369   17436  \n",
       "3      MIR1302-2HG          1   29774   35418  \n",
       "4        MIR1302-2          1   30366   30503  \n",
       "...            ...        ...     ...     ...  \n",
       "68320       TSPY10          Y  752540  755334  \n",
       "68321      TSPY15P          Y  772716  775527  \n",
       "68322      RBMY3AP          Y  835375  839807  \n",
       "68323      TSPY25P          Y  848837  851006  \n",
       "68324      TSPY16P          Y  864903  865415  \n",
       "\n",
       "[68325 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"../Data/ncbi_dataset.tsv\"\n",
    "gene_df = pd.read_csv(file_path, sep=\"\\t\")\n",
    "gene_df = gene_df[[\"Gene ID\", \"Name\", \"Symbol\", \"Chromosome\", \"Begin\", \"End\"]]\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed file saved to ../Data/refined_processed_HiC/K_processed.csv.gz\n",
      "Processed file saved to ../Data/refined_processed_HiC/IMR_processed.csv.gz\n",
      "Processed file saved to ../Data/refined_processed_HiC/GM_processed.csv.gz\n"
     ]
    }
   ],
   "source": [
    "data_dir = \"../Data/all_processed_HiC\"\n",
    "out_put_dir = \"../Data/refined_processed_HiC\"\n",
    "\n",
    "\n",
    "files = [f for f in os.listdir(data_dir) if f.endswith(\".csv.gz\")]\n",
    "\n",
    "for file in files:\n",
    "    # cell line name\n",
    "    cell_line = file.split(\"_\")[0]\n",
    "\n",
    "    file_path = os.path.join(data_dir, file)\n",
    "    df = pd.read_csv(file_path)\n",
    "\n",
    "    # get the average value of 'fq' and 'fdr' for each group\n",
    "    df = df.groupby([\"chr\", \"ibp\", \"jbp\"], as_index=False).agg(\n",
    "        {\"fq\": \"mean\", \"fdr\": \"mean\", \"rawc\": \"first\"}\n",
    "    )\n",
    "\n",
    "    df[\"cell_line\"] = cell_line\n",
    "\n",
    "    # save the processed HiC file\n",
    "    processed_file_path = os.path.join(out_put_dir, f\"{cell_line}_processed.csv.gz\")\n",
    "    df.to_csv(processed_file_path, index=False, compression=\"gzip\")\n",
    "    print(f\"Processed file saved to {processed_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            chr        ibp        jbp   fq  fdr        rawc cell_line\n",
      "121150427  chrX  154055000  154060000  1.0  1.0   98.492010       IMR\n",
      "121150428  chrX  154055000  154065000  1.0  1.0   59.830734       IMR\n",
      "121150429  chrX  154060000  154060000  1.0  1.0  187.421430       IMR\n",
      "121150430  chrX  154060000  154065000  1.0  1.0  102.419624       IMR\n",
      "121150431  chrX  154065000  154065000  1.0  1.0  198.546620       IMR\n"
     ]
    }
   ],
   "source": [
    "file_path = \"../Data/refined_processed_HiC/IMR_processed.csv.gz\"\n",
    "IMR_df = pd.read_csv(file_path)\n",
    "print(IMR_df.tail())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Get example data from the refined csv.gz, get 3 chrs from each file, these chrs exist in each file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/pm/_db3z8953vb1gtk99jkk_sb80000gn/T/ipykernel_75848/2276497908.py:34: DeprecationWarning: Sampling from a set deprecated\n",
      "since Python 3.9 and will be removed in a subsequent version.\n",
      "  sampled_chromosomes = random.sample(common_chromosomes, sample_count)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sampled chromosomes: ['chr17', 'chr12']\n",
      "Processed and saved IMR_processed.csv.gz with 2 common 'chr' groups to ../Example_Data/refined_processed_HiC\n",
      "Processed and saved GM_processed.csv.gz with 2 common 'chr' groups to ../Example_Data/refined_processed_HiC\n",
      "Processed and saved K_processed.csv.gz with 2 common 'chr' groups to ../Example_Data/refined_processed_HiC\n"
     ]
    }
   ],
   "source": [
    "def process_and_sample_data(input_folder, output_folder, sample_count=2):\n",
    "    # Ensure output folder exists\n",
    "    os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "    # Collect the intersection of 'chr' values across all files\n",
    "    common_chromosomes = None\n",
    "\n",
    "    # First pass: Identify common 'chr' values\n",
    "    for filename in os.listdir(input_folder):\n",
    "        if filename.endswith(\".csv.gz\"):\n",
    "            input_path = os.path.join(input_folder, filename)\n",
    "\n",
    "            # Load the compressed CSV file\n",
    "            df = pd.read_csv(input_path, compression=\"gzip\")\n",
    "\n",
    "            # Check if 'chr' column exists\n",
    "            if \"chr\" not in df.columns:\n",
    "                print(f\"'chr' column not found in {filename}. Skipping this file.\")\n",
    "                continue\n",
    "\n",
    "            # Update the set of common chromosomes\n",
    "            file_chromosomes = set(df[\"chr\"].unique())\n",
    "            if common_chromosomes is None:\n",
    "                common_chromosomes = file_chromosomes\n",
    "            else:\n",
    "                common_chromosomes &= file_chromosomes\n",
    "\n",
    "    # Check if we have enough common chromosomes\n",
    "    if common_chromosomes is None or len(common_chromosomes) < sample_count:\n",
    "        print(f\"Not enough common chromosomes across all files. Process aborted.\")\n",
    "        return\n",
    "\n",
    "    # Randomly select the required number of common chromosomes\n",
    "    sampled_chromosomes = random.sample(common_chromosomes, sample_count)\n",
    "    print(f\"Sampled chromosomes: {sampled_chromosomes}\")\n",
    "\n",
    "    # Second pass: Process each file with the sampled chromosomes\n",
    "    for filename in os.listdir(input_folder):\n",
    "        if filename.endswith(\".csv.gz\"):\n",
    "            input_path = os.path.join(input_folder, filename)\n",
    "            output_path = os.path.join(output_folder, filename)\n",
    "\n",
    "            # Load the compressed CSV file\n",
    "            df = pd.read_csv(input_path, compression=\"gzip\")\n",
    "\n",
    "            # Filter the dataframe to include only the selected chromosomes\n",
    "            sampled_data = df[df[\"chr\"].isin(sampled_chromosomes)]\n",
    "\n",
    "            # Save the sampled data to the output file in gzip format\n",
    "            sampled_data.to_csv(output_path, index=False, compression=\"gzip\")\n",
    "            print(f\"Processed and saved {filename} with {sample_count} common 'chr' groups to {output_folder}\")\n",
    "\n",
    "\n",
    "# Define your input and output folders\n",
    "input_folder = \"../Data/refined_processed_HiC\"\n",
    "output_folder = \"../Example_Data/refined_processed_HiC\"\n",
    "\n",
    "# Run the function\n",
    "process_and_sample_data(input_folder, output_folder)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Find each cell line chromosome's range**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed and saved: K_ranges.csv.gz\n",
      "Processed and saved: IMR_ranges.csv.gz\n",
      "Processed and saved: GM_ranges.csv.gz\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from pathos.multiprocessing import ProcessingPool as Pool\n",
    "\n",
    "def process_file(args):\n",
    "    input_folder, filename, output_folder = args\n",
    "    if not filename.endswith(\".csv.gz\"):\n",
    "        return\n",
    "\n",
    "    cell_line = filename.split(\"_\")[0]\n",
    "    file_path = os.path.join(input_folder, filename)\n",
    "    df = pd.read_csv(file_path, compression=\"gzip\")\n",
    "\n",
    "    df[\"ibp_end\"] = df[\"ibp\"] + 5000\n",
    "    df[\"jbp_end\"] = df[\"jbp\"] + 5000\n",
    "\n",
    "    # batch processing\n",
    "    ibp_ranges = df[[\"chr\", \"ibp\", \"ibp_end\"]].rename(columns={\"ibp\": \"start_value\", \"ibp_end\": \"end_value\"})\n",
    "    jbp_ranges = df[[\"chr\", \"jbp\", \"jbp_end\"]].rename(columns={\"jbp\": \"start_value\", \"jbp_end\": \"end_value\"})\n",
    "    ranges = pd.concat([ibp_ranges, jbp_ranges])\n",
    "    ranges[\"cell_line\"] = cell_line\n",
    "\n",
    "    ranges = ranges.drop_duplicates().sort_values(by=[\"chr\", \"start_value\", \"end_value\"])\n",
    "\n",
    "    # merge adjacent ranges\n",
    "    merged_results = []\n",
    "    for chr_id, group in ranges.groupby(\"chr\"):\n",
    "        group = group.sort_values(by=\"start_value\").to_dict(orient=\"records\")\n",
    "        merged = [group[0]]\n",
    "\n",
    "        for entry in group[1:]:\n",
    "            last = merged[-1]\n",
    "            if last[\"end_value\"] >= entry[\"start_value\"]:\n",
    "                last[\"end_value\"] = max(last[\"end_value\"], entry[\"end_value\"])\n",
    "            else:\n",
    "                merged.append(entry)\n",
    "\n",
    "        for entry in merged:\n",
    "            merged_results.append({\n",
    "                \"cell_line\": cell_line,\n",
    "                \"chrID\": chr_id,\n",
    "                \"start_value\": entry[\"start_value\"],\n",
    "                \"end_value\": entry[\"end_value\"],\n",
    "            })\n",
    "\n",
    "    result_df = pd.DataFrame(merged_results, columns=[\"cell_line\", \"chrID\", \"start_value\", \"end_value\"])\n",
    "\n",
    "    # save\n",
    "    output_filename = f\"{cell_line}_ranges.csv.gz\"\n",
    "    output_path = os.path.join(output_folder, output_filename)\n",
    "    result_df.to_csv(output_path, index=False, compression=\"gzip\")\n",
    "    print(f\"Processed and saved: {output_filename}\")\n",
    "\n",
    "def main(input_folder, output_folder, num_workers=4):\n",
    "    files = [(input_folder, filename, output_folder) for filename in os.listdir(input_folder) if filename.endswith(\".csv.gz\")]\n",
    "\n",
    "    # process files in parallel\n",
    "    with Pool(num_workers) as pool:\n",
    "        pool.map(process_file, files)\n",
    "\n",
    "input_folder = \"../Data/refined_processed_HiC\"\n",
    "output_folder = \"../Data/seqs\"\n",
    "\n",
    "\n",
    "main(input_folder, output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total rows: 44\n",
      "   cell_line  chrID  start_value  end_value\n",
      "0          K  chr12       315000    8410000\n",
      "1          K  chr12      8415000    8425000\n",
      "2          K  chr12      8435000    9515000\n",
      "3          K  chr12      9530000    9555000\n",
      "4          K  chr12      9565000   11360000\n",
      "5          K  chr12     11370000   11375000\n",
      "6          K  chr12     11390000   34510000\n",
      "7          K  chr12     39180000  132600000\n",
      "8          K  chr17       605000   12450000\n",
      "9          K  chr17     12455000   18460000\n",
      "10         K  chr17     18465000   19155000\n",
      "11         K  chr17     19160000   19165000\n",
      "12         K  chr17     19180000   19190000\n",
      "13         K  chr17     19205000   19215000\n",
      "14         K  chr17     19220000   21800000\n",
      "15         K  chr17     21810000   21905000\n",
      "16         K  chr17     21920000   21925000\n",
      "17         K  chr17     21930000   21955000\n",
      "18         K  chr17     21965000   21995000\n",
      "19         K  chr17     22040000   22520000\n",
      "20         K  chr17     27085000   36380000\n",
      "21         K  chr17     36385000   37925000\n",
      "22         K  chr17     37930000   37935000\n",
      "23         K  chr17     37940000   37980000\n",
      "24         K  chr17     37995000   38000000\n",
      "25         K  chr17     38010000   38065000\n",
      "26         K  chr17     38070000   38095000\n",
      "27         K  chr17     38100000   38105000\n",
      "28         K  chr17     38115000   38180000\n",
      "29         K  chr17     38185000   46325000\n",
      "30         K  chr17     46330000   46335000\n",
      "31         K  chr17     46340000   46375000\n",
      "32         K  chr17     46380000   46430000\n",
      "33         K  chr17     46435000   46440000\n",
      "34         K  chr17     46445000   46450000\n",
      "35         K  chr17     46455000   46465000\n",
      "36         K  chr17     46470000   46480000\n",
      "37         K  chr17     46485000   46510000\n",
      "38         K  chr17     46515000   46520000\n",
      "39         K  chr17     46525000   46560000\n",
      "40         K  chr17     46565000   46650000\n",
      "41         K  chr17     46655000   70460000\n",
      "42         K  chr17     70465000   81745000\n",
      "43         K  chr17     81790000   82880000\n",
      "Rows where start_value == end_value: 0\n",
      "No rows where start_value == end_value found.\n"
     ]
    }
   ],
   "source": [
    "file_path = \"../Example_Data/seqs/K_ranges.csv.gz\"\n",
    "IMR_df = pd.read_csv(file_path)\n",
    "\n",
    "print(f\"Total rows: {len(IMR_df)}\")\n",
    "print(IMR_df)\n",
    "\n",
    "equal_values_df = IMR_df[IMR_df[\"start_value\"] == IMR_df[\"end_value\"]]\n",
    "\n",
    "print(f\"Rows where start_value == end_value: {len(equal_values_df)}\")\n",
    "\n",
    "if len(equal_values_df) > 0:\n",
    "    print(\"Rows where start_value == end_value:\")\n",
    "    print(equal_values_df)\n",
    "else:\n",
    "    print(\"No rows where start_value == end_value found.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "chr",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ibp",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "jbp",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "fq",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "fdr",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "rawc",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "cell_line",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "9784aff9-c13a-4bba-a601-8bf202c7951d",
       "rows": [
        [
         "0",
         "chr1",
         "770000",
         "770000",
         "1.0",
         "1.0",
         "274.20648",
         "K"
        ],
        [
         "1",
         "chr1",
         "770000",
         "775000",
         "1.0",
         "1.0",
         "184.122",
         "K"
        ],
        [
         "2",
         "chr1",
         "770000",
         "780000",
         "1.00000000000001",
         "0.0",
         "42.198",
         "K"
        ],
        [
         "3",
         "chr1",
         "770000",
         "785000",
         "1.00000000000001",
         "0.0",
         "55.489918",
         "K"
        ],
        [
         "4",
         "chr1",
         "770000",
         "790000",
         "1.00000000000001",
         "0.0",
         "58.17174",
         "K"
        ]
       ],
       "shape": {
        "columns": 7,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>ibp</th>\n",
       "      <th>jbp</th>\n",
       "      <th>fq</th>\n",
       "      <th>fdr</th>\n",
       "      <th>rawc</th>\n",
       "      <th>cell_line</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>770000</td>\n",
       "      <td>770000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>274.206480</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>770000</td>\n",
       "      <td>775000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>184.122000</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>770000</td>\n",
       "      <td>780000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.198000</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>770000</td>\n",
       "      <td>785000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>55.489918</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>770000</td>\n",
       "      <td>790000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>58.171740</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr     ibp     jbp   fq  fdr        rawc cell_line\n",
       "0  chr1  770000  770000  1.0  1.0  274.206480         K\n",
       "1  chr1  770000  775000  1.0  1.0  184.122000         K\n",
       "2  chr1  770000  780000  1.0  0.0   42.198000         K\n",
       "3  chr1  770000  785000  1.0  0.0   55.489918         K\n",
       "4  chr1  770000  790000  1.0  0.0   58.171740         K"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# file_path = \"../Backend/K_chr1_770000_2705000.csv.gz\"\n",
    "# IMR_df = pd.read_csv(file_path)\n",
    "# print(IMR_df.tail())\n",
    "file_path = \"../Data/refined_processed_HiC/K_processed.csv.gz\"\n",
    "IMR_df = pd.read_csv(file_path)\n",
    "IMR_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "VDS",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
