{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import requests\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Generate human chromosome size data**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome sizes saved to /Users/siyuanzhao/Documents/GitHub/CS522_Project/Scripts/chromosome_sizes.txt\n"
     ]
    }
   ],
   "source": [
    "def fetch_chrom_sizes(genome, output_file):\n",
    "    \"\"\"Utility for downloading chromosome sizes from UCSC.\"\"\"\n",
    "    full_file_path = os.path.join(os.getcwd(), output_file)\n",
    "\n",
    "    url = f\"http://hgdownload.soe.ucsc.edu/goldenPath/{genome}/bigZips/{genome}.chrom.sizes\"\n",
    "    response = requests.get(url)\n",
    "\n",
    "    if response.status_code == 200:\n",
    "        with open(full_file_path, \"w\") as f:\n",
    "            for line in response.text.splitlines():\n",
    "                chrom, size = line.split()\n",
    "                f.write(f\"{chrom}\\t{size}\\n\")\n",
    "        print(f\"Chromosome sizes saved to {full_file_path}\")\n",
    "    else:\n",
    "        print(f\"Failed to fetch chromosome sizes for {genome}\")\n",
    "\n",
    "\n",
    "output_file = \"chromosome_sizes.txt\"\n",
    "fetch_chrom_sizes(\"hg38\", output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Generate folding input files**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "def make_dir(d):\n",
    "    \"\"\"Utility for making a directory if not existing.\"\"\"\n",
    "    if not os.path.exists(d):\n",
    "        os.makedirs(d)\n",
    "\n",
    "def get_spe_inter(hic_data, alpha=0.05):\n",
    "    \"\"\"Filter Hi-C data for significant interactions based on the alpha threshold.\"\"\"\n",
    "    hic_spe = hic_data.loc[hic_data['fdr'] < alpha]\n",
    "    return hic_spe\n",
    "\n",
    "def get_fold_inputs(spe_df):\n",
    "    \"\"\"Prepare folding input file from the filtered significant interactions.\"\"\"\n",
    "    spe_out_df = spe_df[['ibp', 'jbp', 'fq', 'chr', 'fdr']]\n",
    "    spe_out_df['w'] = 1\n",
    "    result = spe_out_df[['chr', 'ibp', 'jbp', 'fq', 'w']]\n",
    "    return result\n",
    "\n",
    "def process_hic_files(input_folder, seqs_folder, output_folder, alpha=0.05):\n",
    "    \"\"\"Process Hi-C files by matching with seqs files for reference data and save results in the output folder.\"\"\"\n",
    "    \n",
    "    make_dir(output_folder)\n",
    "    \n",
    "    # Iterate through each file in the Hi-C input folder\n",
    "    for hic_file_name in os.listdir(input_folder):\n",
    "        if hic_file_name.endswith(\".csv.gz\"):\n",
    "            key_word = hic_file_name.split(\"_\")[0]  # Assume keyword is the prefix before the first underscore\n",
    "            \n",
    "            # Find the matching reference file in the seqs folder\n",
    "            seq_file_name = f\"{key_word}_ranges.csv.gz\"\n",
    "            seq_file_path = os.path.join(seqs_folder, seq_file_name)\n",
    "            \n",
    "            if not os.path.exists(seq_file_path):\n",
    "                print(f\"No matching reference file found for {hic_file_name}\")\n",
    "                continue\n",
    "            \n",
    "            # Load Hi-C data and the corresponding reference data\n",
    "            hic_file_path = os.path.join(input_folder, hic_file_name)\n",
    "            all_hic = pd.read_csv(hic_file_path)\n",
    "            spe_hic = get_spe_inter(all_hic, alpha)\n",
    "            reference_df = pd.read_csv(seq_file_path, usecols=[\"chrID\", \"start_value\", \"end_value\"])\n",
    "\n",
    "            # Process each row in the reference file for filtering\n",
    "            for _, row in reference_df.iterrows():\n",
    "                chrID = row[\"chrID\"]\n",
    "                start_value = row[\"start_value\"]\n",
    "                end_value = row[\"end_value\"]\n",
    "                \n",
    "                # Filter Hi-C data based on chrID and ibp range\n",
    "                chr_hic_data = spe_hic[\n",
    "                    (spe_hic[\"chr\"] == chrID) &\n",
    "                    (spe_hic[\"ibp\"] >= start_value) &\n",
    "                    (spe_hic[\"ibp\"] <= end_value)\n",
    "                ]\n",
    "                \n",
    "                if chr_hic_data.empty:\n",
    "                    continue\n",
    "\n",
    "                # Prepare folding input data\n",
    "                fold_hic = get_fold_inputs(chr_hic_data)\n",
    "\n",
    "                # Generate output file name and save\n",
    "                output_file_name = f\"{key_word}.{chrID}.{start_value}.{end_value}.txt\"\n",
    "                fold_hic_path = os.path.join(output_folder, output_file_name)\n",
    "\n",
    "                fold_hic.to_csv(fold_hic_path, header=None, index=None, sep=\"\\t\", mode=\"a\")\n",
    "\n",
    "input_folder = '../Data/refined_processed_HiC'\n",
    "seqs_folder = '../Data/seqs'\n",
    "output_folder = '../Data/Folding_input'\n",
    "process_hic_files(input_folder, seqs_folder, output_folder)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Epigenetic tracks**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>signalValue</th>\n",
       "      <th>pValue</th>\n",
       "      <th>qValue</th>\n",
       "      <th>peak</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, start, end, name, score, strand, signalValue, pValue, qValue, peak]\n",
       "Index: []"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with gzip.open(\"../Data/epigenetic_tracks/GM_H3K27me3.bed.gz\", \"rt\") as f:\n",
    "    df = pd.read_csv(f, sep=\"\\t\", header=None)\n",
    "    df.columns = [\"chrom\", \"start\", \"end\", \"name\", \"score\", \"strand\", \"signalValue\", \"pValue\", \"qValue\", \"peak\"]\n",
    "\n",
    "filtered_df = df[\n",
    "    (df[\"chrom\"] == \"chr8\") &\n",
    "    (df[\"start\"] >= 127150000) &\n",
    "    (df[\"end\"] <= 128260000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Begin</th>\n",
       "      <th>End</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=gene-MIR6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>29774</td>\n",
       "      <td>35418</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=gene-MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=gene-MIR1302-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>34611</td>\n",
       "      <td>36081</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=gene-FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=gene-OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47527</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=gene-ND6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47528</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=gene-TRNE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47529</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=gene-CYTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47530</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>ID=gene-TRNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47531</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>ID=gene-TRNP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47532 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Begin    End Orientation               Symbol\n",
       "0      NC_000001.11  17369  17436           -    ID=gene-MIR6859-1\n",
       "1      NC_000001.11  29774  35418           +  ID=gene-MIR1302-2HG\n",
       "2      NC_000001.11  30366  30503           +    ID=gene-MIR1302-2\n",
       "3      NC_000001.11  34611  36081           -      ID=gene-FAM138A\n",
       "4      NC_000001.11  65419  71585           +        ID=gene-OR4F5\n",
       "...             ...    ...    ...         ...                  ...\n",
       "47527   NC_012920.1  14149  14673           -          ID=gene-ND6\n",
       "47528   NC_012920.1  14674  14742           -         ID=gene-TRNE\n",
       "47529   NC_012920.1  14747  15887           +         ID=gene-CYTB\n",
       "47530   NC_012920.1  15888  15953           +         ID=gene-TRNT\n",
       "47531   NC_012920.1  15956  16023           -         ID=gene-TRNP\n",
       "\n",
       "[47532 rows x 5 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('human_gene_locations_with_names.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('human_gene_locations_with_names.csv')\n",
    "\n",
    "df = df[df['Chromosome'].str.startswith(\"NC_\")].copy()\n",
    "\n",
    "df['Chromosome'] = df['Chromosome'].apply(lambda x: str(int(x.split('.')[0][-2:])))\n",
    "\n",
    "# + -> plus, - -> minus\n",
    "df['Orientation'] = df['Orientation'].map({'+': 'plus', '-': 'minus'})\n",
    "\n",
    "# delete the prefix \"ID=gene-\"\n",
    "df['Symbol'] = df['Symbol'].str.replace(\"ID=gene-\", \"\", regex=False)\n",
    "\n",
    "\n",
    "df.to_csv('output.csv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Begin</th>\n",
       "      <th>End</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>minus</td>\n",
       "      <td>MIR6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>29774</td>\n",
       "      <td>35418</td>\n",
       "      <td>plus</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>plus</td>\n",
       "      <td>MIR1302-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>34611</td>\n",
       "      <td>36081</td>\n",
       "      <td>minus</td>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>plus</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41706</th>\n",
       "      <td>20</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>minus</td>\n",
       "      <td>ND6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41707</th>\n",
       "      <td>20</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>minus</td>\n",
       "      <td>TRNE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41708</th>\n",
       "      <td>20</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>plus</td>\n",
       "      <td>CYTB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41709</th>\n",
       "      <td>20</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>plus</td>\n",
       "      <td>TRNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41710</th>\n",
       "      <td>20</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>minus</td>\n",
       "      <td>TRNP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41711 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  Begin    End Orientation       Symbol\n",
       "0               1  17369  17436       minus    MIR6859-1\n",
       "1               1  29774  35418        plus  MIR1302-2HG\n",
       "2               1  30366  30503        plus    MIR1302-2\n",
       "3               1  34611  36081       minus      FAM138A\n",
       "4               1  65419  71585        plus        OR4F5\n",
       "...           ...    ...    ...         ...          ...\n",
       "41706          20  14149  14673       minus          ND6\n",
       "41707          20  14674  14742       minus         TRNE\n",
       "41708          20  14747  15887        plus         CYTB\n",
       "41709          20  15888  15953        plus         TRNT\n",
       "41710          20  15956  16023       minus         TRNP\n",
       "\n",
       "[41711 rows x 5 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../Data/gene_list.csv', sep='\\t')\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "VDS",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
